### PR TITLE
Adjust error code and epochtime

### PR DESCRIPTION
**Pull request by greatect**
* Define all app error codes explicitly
* Reallocate jobs in isTxValid and executeTx
* Use the appropriate epoch time 
* Fix fixup in config

### CODE REVIEW

File: state.go [hunk 1]
```diff
@@ -23,6 +23,11 @@ import (
 const genesisEpoch uint64 = 1
 const epochInterval int64 = 5
 
+var (
+	errDocDescriptor = errors.New("insufficient descriptors uploaded")
```
Comment (sc0Vu) [inline state.go:27]: Maybe rename to `errDocInsufficientDescriptor`.

File: errors.go [hunk 2]
```diff
@@ -2,28 +2,55 @@ package katzenmint
...
+	ErrTxDocEpoch               = KatzenmintError{Code: 0x15, Msg: "document epoch inconsistent with transaction epoch"}
+	ErrTxDocNotAuthorized       = KatzenmintError{Code: 0x16, Msg: "document is not authorized"}
+	ErrTxAuthorityParse         = KatzenmintError{Code: 0x17, Msg: "cannot parse authority"}
+	ErrTxAuthorityNotAuthorized = KatzenmintError{Code: 0x18, Msg: "descriptor is not authorized"}
+	ErrTxCommandNotFound        = KatzenmintError{Code: 0x19, Msg: "transaction command not found"}
 
-	ErrTxNonAuthorized   = KatzenmintError{Msg: "non authorized authority", Code: 0x31}
-	ErrTxCommandNotFound = KatzenmintError{Msg: "transaction command not found", Code: 0x32}
+	// Transaction Execution Errors
+	ErrTxExeEpoch   = KatzenmintError{Code: 0x21, Msg: "expect transaction epoch within +-1 to current epoch"}
```
Comment (sc0Vu) [inline errors.go:39]: Maybe rename to `ErrWrongEpoch` or others.

File: app.go [hunk 3]
```diff
@@ -82,22 +83,47 @@ func (app *KatzenmintApplication) isTxValid(rawTx []byte) (tx *Transaction, err
...
+		}
+
 	case AddConsensusDocument:
+		payload = []byte(tx.Payload)
+		doc, err = s11n.VerifyAndParseDocument(payload)
+		if err != nil {
+			err = ErrTxDocFalseVerification
+			return
+		}
+		if doc.Epoch != tx.Epoch {
```
Comment (sc0Vu) [inline app.go:109]: Should we remove Epoch in tx? Thought this might be redundant.

File: errors.go [hunk 4]
```diff
@@ -2,28 +2,55 @@ package katzenmint
...
+	ErrTxWrongSignature     = KatzenmintError{Code: 0x04, Msg: "wrong signature in transaction"}
 
-	ErrTxDescInvalidVerifier   = KatzenmintError{Msg: "cannot get descriptor verifier", Code: 0x11}
-	ErrTxDescFalseVerification = KatzenmintError{Msg: "cannot verify and parse descriptor", Code: 0x12}
+	// Transaction Specific Errors
+	ErrTxDescInvalidVerifier    = KatzenmintError{Code: 0x11, Msg: "cannot get descriptor verifier"}
+	ErrTxDescFalseVerification  = KatzenmintError{Code: 0x12, Msg: "cannot verify and parse descriptor"}
+	ErrTxDescNotAuthorized      = KatzenmintError{Code: 0x13, Msg: "authority is not authorized"}
+	ErrTxDocFalseVerification   = KatzenmintError{Code: 0x14, Msg: "cannot verify and parse document"}
+	ErrTxDocEpoch               = KatzenmintError{Code: 0x15, Msg: "document epoch inconsistent with transaction epoch"}
```
Comment (sc0Vu) [inline errors.go:32]: Maybe rename to `ErrTxDocEpochNotEqual`

File: state.go [hunk 5]
```diff
@@ -50,7 +55,8 @@ type KatzenmintState struct {
 	validators       map[string]pc.PublicKey
 	validatorUpdates []abcitypes.ValidatorUpdate
 
-	deferCommit []func()
+	deferCommit     []func()
+	prevCommitError error
```
Comment (sc0Vu) [inline state.go:59]: Are we going to use `prevCommitError`, eg: print?

File: state.go [hunk 6]
```diff
@@ -181,7 +188,7 @@ func (state *KatzenmintState) Commit() ([]byte, error) {
 	state.Lock()
 	defer state.Unlock()
 
-	var errDoc error
+	var errDoc error = nil
```
Comment (sc0Vu) [inline state.go:191]: Might better to rename errDoc, also I think default errDoc should be nil.

**Review by sc0Vu — CHANGES_REQUESTED** [review]

File: state.go (hunk 5 above)
Comment (greatect) [reply state.go:59]: Its meant to avoid the app reporting the same error when commit fails over a period

File: app.go (hunk 3 above)
Comment (greatect) [reply app.go:109]: Currently we still need tx.epoch when posting a descriptor. If we put such information in the descriptor structure then I think tx.epoch could be removed.

File: state.go (hunk 1 above)
Comment (greatect) [reply state.go]: Fixed

File: state.go [hunk 7]
```diff
@@ -23,6 +23,11 @@ import (
 const genesisEpoch uint64 = 1
 const epochInterval int64 = 5
 
+var (
+	errDocDescriptor = errors.New("insufficient descriptors uploaded")
+	errDocProvider   = errors.New("no providers uploaded")
```
[the remark this answers is not in the document]
Comment (greatect) [reply state.go]: Fixed

File: errors.go (hunk 2 above)
Comment (greatect) [reply errors.go]: Fixed

File: errors.go (hunk 4 above)
Comment (greatect) [reply errors.go]: Fixed

File: state.go (hunk 6 above)
Comment (greatect) [reply state.go]: Fixed

File: state.go [hunk 8]
```diff
@@ -206,12 +213,22 @@ func (state *KatzenmintState) Commit() ([]byte, error) {
...
-		err = errDoc
+	if err != nil {
+		return nil, err
 	}
 	state.appHash = appHash
 
-	return appHash, err
+	if errDoc == errDocDescriptor || errDoc == errDocProvider {
+		// Report these errors when they first appear
+		if errDoc == state.prevCommitError {
```
[the remark this answers is not in the document]
Comment (greatect) [reply state.go]: Fixed

**Review by sc0Vu — APPROVED** [review]
LGTM

**Pull request merged**